subsys/usb/class/bluetooth.c

```diff
@@ -0,0 +1,236 @@
+/*
+ * Wireless / Bluetooth USB class
+ *
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <init.h>
+
+#include <usb/usb_device.h>
+#include <usb/usb_common.h>
+#include <usb_descriptor.h>
+
+#include <net/buf.h>
+
+#include <bluetooth/buf.h>
+#include <bluetooth/hci_raw.h>
+#include <bluetooth/l2cap.h>
+
+#define SYS_LOG_LEVEL SYS_LOG_LEVEL_WARNING
+#define SYS_LOG_DOMAIN "usb/bluetooth"
+#include <logging/sys_log.h>
+
+#if !defined(CONFIG_USB_COMPOSITE_DEVICE)
+static u8_t interface_data[64];
+#endif
+
+static K_FIFO_DEFINE(rx_queue);
+
+/* HCI command buffers */
+#define CMD_BUF_SIZE BT_BUF_RX_SIZE
+NET_BUF_POOL_DEFINE(tx_pool, CONFIG_BT_HCI_CMD_COUNT, CMD_BUF_SIZE,
+		    sizeof(u8_t), NULL);
+
+#define BT_L2CAP_MTU 64
+/* Data size needed for ACL buffers */
+#define BT_BUF_ACL_SIZE BT_L2CAP_BUF_SIZE(BT_L2CAP_MTU)
+NET_BUF_POOL_DEFINE(acl_tx_pool, 2, BT_BUF_ACL_SIZE, sizeof(u8_t), NULL);
+
+/* Stack for thread */
+static K_THREAD_STACK_DEFINE(bluetooth_thread_stack, 512);
+static struct k_thread bluetooth_thread_data;
+
+static void bluetooth_status_cb(enum usb_dc_status_code status, u8_t *param)
+{
+	/* Check the USB status and do needed action if required */
+	switch (status) {
+	case USB_DC_ERROR:
+		SYS_LOG_DBG("USB device error");
+		break;
+	case USB_DC_RESET:
+		SYS_LOG_DBG("USB device reset detected");
+		break;
+	case USB_DC_CONNECTED:
+		SYS_LOG_DBG("USB device connected");
+		break;
+	case USB_DC_CONFIGURED:
+		SYS_LOG_DBG("USB device configured");
+		break;
+	case USB_DC_DISCONNECTED:
+		SYS_LOG_DBG("USB device disconnected");
+		break;
+	case USB_DC_SUSPEND:
+		SYS_LOG_DBG("USB device suspended");
+		break;
+	case USB_DC_RESUME:
+		SYS_LOG_DBG("USB device resumed");
+		break;
+	case USB_DC_UNKNOWN:
+	default:
+		SYS_LOG_DBG("USB unknown state");
+		break;
+	}
+}
+
+static void bluetooth_thread(void)
+{
+	SYS_LOG_DBG("Start USB Bluetooth thread");
+
+	while (true) {
+		struct net_buf *buf;
+
+		buf = net_buf_get(&rx_queue, K_FOREVER);
+
+		switch (bt_buf_get_type(buf)) {
+		case BT_BUF_EVT:
+			usb_transfer_sync(CONFIG_BLUETOOTH_INT_EP_ADDR,
+					  buf->data, buf->len,
+					  USB_TRANS_WRITE | USB_TRANS_NO_ZLP);
```

Comment:
For some reason without flag `USB_TRANS_NO_ZLP` we always send ZLP after each packet, is this desired behaviour?

Reply:
> For some reason without flag USB_TRANS_NO_ZLP we always send ZLP after each packet, is this desired behaviour?
No this is not expected, we should only send ZLP when payload is a multiple of emps, to mark end of the transfer.

> I understand why ZLP is needed for ECM, since there is no packet length in ECM protocol and we mark end of the packet (and ONLY when packet is 64x bytes long), but why we need it for Bluetooth?
Actually, this is how a transfer is specified, the USB 'core' layer does not take care about payload (ECM, BT...), and needs a way to determine when a transfer is complete.

From USB spec:
"5.8.3 Bulk Transfer Packet Size Constraints
An endpoint must always transmit data payloads with a data field less than or equal to the endpoint’s
reported wMaxPacketSize value. When a bulk IRP involves more data than can fit in one maximum-sized data payload, all data payloads are required to be maximum size except for the last data payload, which will contain the remaining data. A bulk transfer is complete when the endpoint does one of the following:
• Has transferred exactly the amount of data expected
• Transfers a packet with a payload size less than wMaxPacketSize or transfers a zero-length packet"

Reply:
Sorry, my bad, btw I have fixed issue with ZLP in PR #7110